### ISSUE TITLE
MAINT: Tracking existing FutureWarnings

### ISSUE THREAD

**Issue by m-richards** [geopandas/geopandas#2263]
Following on from discussion in #2100, this is a list of all futurewarnings currently in the codebase. Based on the comment,
> I think that anything that has a warning since geopandas 0.8 or older can be turned into an error

I've gone through and annotated each with the git sha they were added in, and the release they were part of (with no errors I hope, but wouldn't surprise me if I've messed a couple up). I think it makes sense to keep this as a meta issue tracking all of these.

- array.py
	- [ ] `exact_equals: ` 1fede1cb - 05/2020 - 0.7
	- [ ] `almost_equals: ` 1fede1cb - 05/2020 - 0.7
- base.py
	- `cascaded_union` befb2e88 - 08/2021 - 0.10 -> deprecate after 0.12?
- geodataframe.py
	- [ ] `crs_mismatch`  (comment TODO: raise error in 0.9 or 0.10.) -> #2100
	- [ ] `__init__` - assigning a crs to a gdf without a geometry column - edf6e964 - 04/2020 -> 0.8
	- [ ] `crs.setter` - assigning a crs to a gdf without a geometry column - edf6e964 - 04/2020 -> 0.8
	- `explode` index parts default - 422b87d3 - 09/2021 -> 0.10 -> deprecate after 0.12?

- geoseries.py
	- [ ] `crs_mismatch`  (comment TODO: raise error in 0.9 or 0.10.): -> #2100
	- [ ] `__new__` - non geometry data to series: 3f8ed034 - 08/2019 - 0.6.2
	- [ ] `explode` index parts default: 422b87d3 - 09/2021 - 0.10 -> deprecate after 0.12?
- plotting.py
	- [ ] `plot_series`/`plot_dataframe` 'colormap' is deprecated, please use 'cmap' instead: 7bc3166c - 09/2019 - 0.6.2
	- [ ] `plot_series`/`plot_dataframe` 'axes' is deprecated, please use 'ax' instead: 7bc3166c - 09/2019 - 0.6.2
- sindex.py
	- `SpatialIndex.__init__`: af867de8 - 02/2021 - 0.9 -> deprecate after 0.11?
	- `RTreeIndex.nearest` undocumented behaviour change: 64c5813e 09/2021 - 0.10
- tools/crs.py
	- [ ] `explicit_crs_from_epsg`: 1d212b2a 01/2020 - 0.6.3 
	- [ ] `epsg_from_crs`: 1d212b2a 01/2020 - 0.6.3
	- [ ] `get_epsg_file_contents`: 1d212b2a 01/2020 - 0.6.3
- tools/sjoin.py
	- `sjoin` op -> predicate language change: 8f08539b 09/2021 - 0.10
	- [ ] `_geom_predicate_query` spatial index is empty 72c57c0e 06/2020 - 0.8.1 - filterwarning, need to check what this is in reference to

**Comment by m-richards** [comment]
There is also a set of Deprecation warnings that aren't covered in the above list.

* Set operation overloads `__xor__, __or__, __and__, __sub__` in geoseries.py/ geodataframe.py: f13dafb - 04/20 - 0.8
* io
     * file.py `read_file` and `to_file` top level functions: 535dcad5 - 5/20 - 0.8
     * sql.py `read_postgis`: 535dcad5 - 5/20 - 0.8
 * plotting.py
     * `plot_point_collection`, `plot_point_collection`, `plot_polygon_collection`: 708b38ea - 4/20 - 0.8

The subtleties of FutureWarning vs DeprecationWarning are a little bit lost on me, I understand there's a difference in intended audience and urgency, but not sure what that actually translates to in terms of a deprecation period, so this is more intended as reference than anything else right now.

**Comment by martinfleis** [comment]
> The subtleties of FutureWarning vs DeprecationWarning are a little bit lost on me, I understand there's a difference in intended audience and urgency, but not sure what that actually translates to in terms of a deprecation period, so this is more intended as reference than anything else right now.

I don't think we follow any strict rules here. I usually tend to use FutureWarning in projects like GeoPandas where I assume that end user uses the most of methods.

**Comment by jorisvandenbossche** [comment]
In general, FutureWarnings are more "noisy" (always shown by default), while DeprecationWarnings are by default only shown if you directly use something that raises them (but not if you use a function that uses a deprecated function). Of course for package development, it is still recommended that you set DeprecationWarning to always be shown in eg your tests.

So for pandas we have the general rule: we use FutureWarning for most user-facing deprecations, unless we think it is something really widely used by downstream packages as well, (or semi-private functionality of which we assume some packages will use it) and then we first use a DeprecationWarning before switching to FutureWarning (to give those packages the time to fix the warnings, before their users see them as well).

Long story short: I think we can certainly change those DeprecationWarnings into FutureWarnings now (since they already exist for some time)

**Comment by m-richards** [comment]
I decided to have a look at this again (the list probably needs an update, but can start with what is there) and was wondering what makes sense to change for 1.0. I think 
* explode `index_parts` default is a good candidate, since it's breaking and we can clean up warnings.
* op -> predicate language change would be good to do - I'm just not sure how best to handle, do we remove any code handling `op` entirely, or do we raise if `op` is given?
* `_geom_predicate_query` is a filter warning in our own code for a warning dropped in #1444 so is safe to remove.

Edit: All the ones listed as deprecation warnings were converted in #2447 so I think these are safe to remove as well.

Edit2: So that the above comment stays in context, here are all the additional futurewarnings:
* [x] geom_almost_equals https://github.com/geopandas/geopandas/commit/5ebf9488b653e6ccb65386d9c75861864a667927 -> 0.14
* [x] geoparquet version  (and feather) -> schema_version a844ee9 -> 0.11.1, I suppose we should do this one too
* [ ] unary union -> union all a1ad6816bae -> main/ 1.0
* [ ] seed -> rng in sample_points 969f2248 -> main/ 1.0 ->#3631
* [ ] crs setter (https://github.com/geopandas/geopandas/pull/3088) -> 1.0 -> futurewarning in #3631 -> remove in 1.4
* [ ] `use_pygeos` option (https://github.com/geopandas/geopandas/pull/3283) -> 1.0 -> #3631
* [ ] ``include_fields`` and ``ignore_fields`` keywords in ``read_file()`` (#3133) -> 1.0
* [ ] The `drop` keyword in `set_geometry` (#3237) -> 1.0

**Comment by martinfleis** [comment]
I would  personally remove all deprecation and future warnings that are not new in 1.0 and switch the behaviour as was advertised.

> do we remove any code handling op entirely, or do we raise if op is given

I'd remove it.

**Comment by m-richards** [comment]
> > 'colormap' is deprecated, please use 'cmap' instead [...] 'axes' is deprecated, please use 'ax' instead
> 
> What about these deprecated in 2015? I think we should either remove those options or remove the warning at this point.

That's a fair point, no sense in having a permanent warning. If we're making changes for these cases we may as well decide to remove them an clean up the code.

**Comment by jorisvandenbossche** [comment]
I opened https://github.com/geopandas/geopandas/pull/3334 for the parquet/feather `version` keyword. I think everything else in the list above is for later, so moving this to the next milestone
